rules/solid/src/Rector/Variable/MoveVariableDeclarationNearReferenceRector.php

```diff
@@ -0,0 +1,115 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\SOLID\Rector\Variable;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\Assign;
+use PhpParser\Node\Expr\Variable;
+use Rector\Core\Rector\AbstractRector;
+use Rector\Core\RectorDefinition\CodeSample;
+use Rector\Core\RectorDefinition\RectorDefinition;
+use Rector\NodeNestingScope\ParentScopeFinder;
+use Rector\NodeTypeResolver\Node\AttributeKey;
+
+/**
+ * @see \Rector\SOLID\Tests\Rector\Variable\MoveVariableDeclarationNearReferenceRector\MoveVariableDeclarationNearReferenceRectorTest
+ */
+final class MoveVariableDeclarationNearReferenceRector extends AbstractRector
+{
+    /** @var ParentScopeFinder */
+    private $parentScopeFinder;
+
+    public function __construct(ParentScopeFinder $parentScopeFinder)
+    {
+        $this->parentScopeFinder = $parentScopeFinder;
+    }
+
+    public function getDefinition(): RectorDefinition
+    {
+        return new RectorDefinition('Move variable declaration near its reference', [
+            new CodeSample(
+                <<<'CODE_SAMPLE'
+$var = 1;
+if ($condition === null) {
+    return $var;
+}
+CODE_SAMPLE
+                ,
+                <<<'CODE_SAMPLE'
+if ($condition === null) {
+    $var = 1;
+    return $var;
+}
+CODE_SAMPLE
+            ),
+        ]);
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getNodeTypes(): array
+    {
+        return [Assign::class];
+    }
+
+    /**
+     * @param Assign $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        $assign = $node;
+        $variable = $node->var;
+        if (!$variable instanceof Variable) {
+            return null;
+        }
+
+        $parentScope = $this->parentScopeFinder->find($assign);
+        if ($parentScope === null) {
+            return null;
+        }
+
+        $usage = $this->findFirstVariableUsageInScope($variable, $assign, $parentScope);
+        if ($usage === null) {
+            return null;
+        }
+
+        $usageStatement = $usage->getAttribute(AttributeKey::CURRENT_STATEMENT);
+        $assignStatement = $assign->getAttribute(AttributeKey::CURRENT_STATEMENT);
+
+        $this->addNodeBeforeNode($assignStatement, $usageStatement);
+        $this->removeNode($assignStatement);
```

Comment:
@TomasVotruba This node manipulation seems to have no effect in case `$parentScope` is a `Closure`. Everything else looks good in the debugger. Perhaps a bug in Rector?

Reply:
Some nodes cannot be removed just like that.

E.g. if you remove `$if->cond` from `if($someCond)`, the `If_` would collapse.

In non-simple cases like these, one needs to work/remove the propper node, e.g. `$if->cond = null` or `$this->remove($ifNode)`.

Reply:
Right, but that doesn't change the fact that this piece of code works fine on every scope except a closure. The code being manipulated is the same in every other way, so something about closures seem to cause unexpected problem.

Reply:
Agreed. The Closure needs to be handled in custom way here

Reply:
I really don't see how to do that if node changes to a closure aren't registered properly in Rector. Is there some sort of weird exclusion in some collectors or visitors somewhere? I don't have enough Rector knowledge to be able to debug something like this. `$this->removeNode` does nothing.

Reply:
It's hard to say without real code. No need to work with `removeNode()`, as it's designed only for simple cases with 1 way to solve. In your case there might be many ways you'd like to solve it.

In general, I'd look for a `Closure_` and its location file, e.g.

```php
$parent = $assign->getAttribute(AttributeKey::PARENT);
if ($parent instanceof Closure_) {
   $closureParent = $parent->getAttribute(AttributeKey::PARENT);

   // $closureParent - what type is it? is it argument? assign? 
   // there are various way to handle this; remove `$closureParent` or null one of $closureParent properties, etc.
}
```

Real minimal PHP code with the closure and the node you're trying to remove would help, e.g.

```php
 $value = function () { 
-   $item = 5;
 };
```

Reply:
The test case is in the PR already: rules/solid/tests/Rector/Variable/MoveVariableDeclarationNearReferenceRector/Fixture/variable_in_closure.php.inc

Reply:
Thanks!

Link to the file would save me few clicks. Moreover on phone, where this takes copy-pasting pro skills

Reply:
![image](https://user-images.githubusercontent.com/924196/96038739-e2894780-0e67-11eb-866a-a153844b6934.png)


So if you the removing works correctly. The adding node is broken :)